### PR TITLE
(SIMP-4959) Add JSON schema for compliance data

### PR DESCRIPTION
**Pull request by jeefberkey**
This commit adds:
* A JSON schema for JSON files in the data/compliance_profiles directory
* A rake task to validate the data against the schema

SIMP-4959 #close

### CODE REVIEW

File: data/compliance_profiles/schema.json [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+        }
+      },
+      "patternProperties": {
+        "^((?!version).)*$": {
+          "type": "object",
+          "additionalProperties": false,
+          "patternProperties": {
+            "^\\w+(?:::\\w+)+$": {
+              "type": "object",
+              "properties": {
```
Comment (jeefberkey) [inline data/compliance_profiles/schema.json]: @heliocentric Can you make sure that the properties listed here contain all the possibilities that the mapper can handle?

File: data/compliance_profiles/schema.md [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+### `compliance_map_name`
+
+Name of the complince standard that is being mapped
+
+#### `puppet::class::parameter`
+
+The fully-namespaced Puppet class parameter, as it would be written if you were
+to set it in hiera.
+
+##### `value`
```
Comment (jeefberkey) [inline data/compliance_profiles/schema.md]: @heliocentric Can you mention here how the value field works when you are managing an array or a hash?

File: data/compliance_profiles/schema.json [hunk 3]
```diff
@@ -0,0 +1,55 @@
+{
+  "$schema": "http://json-schema.org/draft-06/schema#",
```
Comment (op-ct) [inline data/compliance_profiles/schema.json]: @jeefberkey  How do you feel about switching over to [draft-07][d7] and a d7-capable validator?

draft07 was published earlier this year, and looks pretty good.  It didn't add or change the behavior of any keywords, but it introduced features we could probably use, like if-then-else conditionals and comments for schema maintainers.

The downside is that development stopped on the [`json-schema`][g1] gem and  over a year ago, so it doesn't support draft07 .   There's another gem, [`json_schemer`][g2], that does (and is the only ruby implementation currently recognized by [json-schema.org][ji]).  However, it's pretty new and the documentation is sparse.

[ji]: http://json-schema.org/implementations.html#validator-ruby
[d7]: http://json-schema.org/draft-07/json-schema-release-notes.html
[g1]: https://github.com/ruby-json-schema/json-schema
[g2]:  https://github.com/davishmcclurg/json_schemer

File: rakelib/compliance.rake [hunk 4]
```diff
@@ -0,0 +1,30 @@
+require 'json'
+require 'json-schema'
+
+task :validate => 'compliance:schema:validate'
+
+namespace :compliance do
+  namespace :schema do
```
Comment (op-ct) [inline rakelib/compliance.rake]: `compliance:schema:` still seems a little generic for the overall rake task list.

What about something more specific, like `compliance_markup:schema` or `compliance:profile:schema`?

File: data/compliance_profiles/schema.json [hunk 5]
```diff
@@ -0,0 +1,55 @@
+{
```
Comment (op-ct) [inline data/compliance_profiles/schema.json]: Why is the schema in the Puppet module's `data/` directory?

Unless Hiera uses this file for APL, I'd expect it to live in a somewhere else, like `schemas/` or maybe `utils/schemas/`.

File: data/compliance_profiles/schema.json (hunk 3 above)
Comment (op-ct) [reply data/compliance_profiles/schema.json]: **Update:** development on the [`json-schema`][g1] gem just picked up again (new maintainer).

[g1]: https://github.com/ruby-json-schema/json-schema

**Comment by jeefberkey** [comment]
Closing because there's a v2 now

**Pull request closed without merging**